File: app/src/main/java/com/ubergeek42/WeechatAndroid/adapters/ChatLinesAdapter.java
// Copyright 2012 Keith Johnson
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.ubergeek42.WeechatAndroid.adapters;

import androidx.annotation.AnyThread;
import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import com.ubergeek42.WeechatAndroid.R;
import com.ubergeek42.WeechatAndroid.Weechat;
import com.ubergeek42.WeechatAndroid.relay.Buffer;
import com.ubergeek42.WeechatAndroid.relay.BufferEye;
import com.ubergeek42.WeechatAndroid.relay.Line;
import com.ubergeek42.WeechatAndroid.relay.Lines;
import com.ubergeek42.WeechatAndroid.service.P;
import com.ubergeek42.WeechatAndroid.utils.AnimatedRecyclerView;
import com.ubergeek42.WeechatAndroid.utils.CopyPaste;
import com.ubergeek42.WeechatAndroid.utils.LineView;
import com.ubergeek42.WeechatAndroid.utils.Utils;
import com.ubergeek42.cats.Cat;
import com.ubergeek42.cats.Kitty;
import com.ubergeek42.cats.Root;
import com.ubergeek42.weechat.ColorScheme;

import java.util.ArrayList;
import java.util.List;

import static com.ubergeek42.WeechatAndroid.R.layout.more_button;
import static com.ubergeek42.WeechatAndroid.R.layout.read_marker;
import static com.ubergeek42.WeechatAndroid.relay.Buffer.PRIVATE;
import static com.ubergeek42.WeechatAndroid.relay.Lines.HEADER_POINTER;
import static com.ubergeek42.WeechatAndroid.relay.Lines.MARKER_POINTER;

import static com.ubergeek42.WeechatAndroid.utils.Utils.Predicate;
import static com.ubergeek42.WeechatAndroid.utils.Assert.assertThat;


public class ChatLinesAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements BufferEye {
    final private @Root Kitty kitty = Kitty.make("ChatLinesAdapter");

    private final AnimatedRecyclerView uiLines;
    private @Nullable Buffer buffer;
    private List<Line> lines = new ArrayList<>();
    volatile private List<Line> _lines = new ArrayList<>();

    @MainThread public ChatLinesAdapter(AnimatedRecyclerView animatedRecyclerView) {
        this.uiLines = animatedRecyclerView;
        setHasStableIds(true);
    }

    @MainThread public synchronized void setBuffer(@Nullable Buffer buffer) {
        this.buffer = buffer;
        kitty.setPrefix(buffer == null ? null : buffer.shortName);
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    //////////////////////////////////////////////////////////////////////////////////////////////// row
    ////////////////////////////////////////////////////////////////////////////////////////////////

    private static class Row extends RecyclerView.ViewHolder {
        private LineView lineView;

        @MainThread Row(View view) {
            super(view);
            lineView = (LineView) view;
            lineView.setOnLongClickListener(CopyPaste.copyPaste);
        }

        @MainThread void update(Line line) {
            lineView.setTag(line);
            lineView.setText(line);
        }

        @MainThread void cancelAnimation() {
            lineView.cancelAnimation();
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    //////////////////////////////////////////////////////////////////////////////////////////////// read marker
    ////////////////////////////////////////////////////////////////////////////////////////////////

    private static class ReadMarkerRow extends RecyclerView.ViewHolder {
        private View view;

        @MainThread ReadMarkerRow(View view) {
            super(view);
            this.view = view;
        }

        @MainThread void update() {
            view.setBackgroundColor(0xFF000000 | ColorScheme.get().chat_read_marker[0]);
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    //////////////////////////////////////////////////////////////////////////////////////////////// header
    ////////////////////////////////////////////////////////////////////////////////////////////////

    private static class Header extends RecyclerView.ViewHolder implements View.OnClickListener {
        private LineView title;
        private Button button;
        private ChatLinesAdapter adapter;
        private Lines.STATUS status = Lines.STATUS.CAN_FETCH_MORE;

        @MainThread Header(View header, ChatLinesAdapter adapter) {
            super(header);
            this.adapter = adapter;
            title = header.findViewById(R.id.title);
            title.setOnLongClickListener(CopyPaste.copyPaste);
            button = header.findViewById(R.id.button_more);
            button.setOnClickListener(this);
        }

        @MainThread void update() {
            if (adapter.buffer == null) return;
            updateButton();
            updateTitle();
        }

        @MainThread private void updateButton() {
            if (adapter.buffer == null) return;
            final Lines.STATUS s = adapter.buffer.getLinesStatus();
            if (status == s) return;
            status = s;
            if (s == Lines.STATUS.EVERYTHING_FETCHED) {
                button.setVisibility(View.GONE);
            } else {
                button.setVisibility(View.VISIBLE);
                boolean more = s == Lines.STATUS.CAN_FETCH_MORE;
                button.setEnabled(more);
                button.setText(button.getContext().getString(more ? R.string.more_button : R.string.more_button_fetching));
            }
        }

        @MainThread private void updateTitle() {
            if (adapter.buffer == null) return;
            Line titleLine = adapter.buffer.titleLine;
            if (titleLine == null || TextUtils.isEmpty(titleLine.spannable) || !adapter.buffer.linesAreReady()) {
                title.setVisibility(View.GONE);
                return;
            }
            title.setVisibility(View.VISIBLE);
            Utils.setBottomMargin(title, button.getVisibility() == View.GONE ? (int) P._4dp : 0);
            title.setText(titleLine);
            title.setTag(titleLine);
        }

        @MainThread @Override public void onClick(View v) {
            if (adapter.buffer == null) return;
            adapter.buffer.requestMoreLines();
            updateButton();
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    //////////////////////////////////////////////////////////////////////////////////////////////// RecyclerView.Adapter methods
    ////////////////////////////////////////////////////////////////////////////////////////////////

    final private static int HEADER_TYPE = -1, LINE_TYPE = 0, MARKER_TYPE = 1;

    @MainThread @Override public int getItemViewType(int position) {
        long pointer = lines.get(position).pointer;
        if (pointer == HEADER_POINTER) return HEADER_TYPE;
        if (pointer == MARKER_POINTER) return MARKER_TYPE;
        return LINE_TYPE;
    }

    @MainThread @Override public @NonNull RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater i = LayoutInflater.from(parent.getContext());
        if (viewType == HEADER_TYPE) return new Header(i.inflate(more_button, parent, false), this);
        else if (viewType == MARKER_TYPE) return new ReadMarkerRow(i.inflate(read_marker, parent, false));
        else return new Row(new LineView(parent.getContext()));
    }

    @MainThread @Override public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        long pointer = lines.get(position).pointer;
        if (pointer == HEADER_POINTER) ((Header) holder).update();
        else if (pointer == MARKER_POINTER) ((ReadMarkerRow) holder).update();
        else ((Row) holder).update(lines.get(position));
    }

    @MainThread @Override public int getItemCount() {
        return lines.size();
    }

    @MainThread @Override public long getItemId(int position) {
        return lines.get(position).pointer;
    }

    @MainThread @Override public void onViewDetachedFromWindow(@NonNull RecyclerView.ViewHolder holder) {
        if (holder instanceof Row) ((Row) holder).cancelAnimation();
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////////////

    // get new lines, perform a simple diff and dispatch change notifications to RecyclerView
    // this might be called by multiple threads in rapid succession
    // in case non-main thread calls this before the Runnable that sets `lines` is executed,
    // store the new list in `_lines` so that we can produce a proper diff
    @AnyThread private synchronized void onLinesChanged() {
        if (buffer == null) return;
        final ArrayList<Line> newLines = buffer.getLinesCopy();

        final boolean hack = _lines.size() == 1 && newLines.size() > 1;

        final DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffCallback(_lines, newLines), false);
        _lines = newLines;

        Weechat.runOnMainThreadASAP(() -> {
            lines = newLines;
            diffResult.dispatchUpdatesTo(ChatLinesAdapter.this);
            if (uiLines.getOnBottom()) {
                if (hack) uiLines.scrollToPosition(getItemCount() - 1);
                else uiLines.smoothScrollToPosition(getItemCount() - 1);
            }
            else uiLines.flashScrollbar();
            uiLines.scheduleAnimationRestoring();
        });
    }

    @AnyThread private void updateHeader() {
        Weechat.runOnMainThread(() -> notifyItemChanged(0));
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    //////////////////////////////////////////////////////////////////////////////////////////////// BufferEye
    ////////////////////////////////////////////////////////////////////////////////////////////////

    // increasing `style` will make all ViewHolders update visual characteristics
    @MainThread @Override public synchronized void onGlobalPreferencesChanged(boolean numberChanged) {
        if (numberChanged && buffer != null) {
            onLinesChanged();
        } else {
            notifyItemRangeChanged(0, _lines.size());
        }
    }

    @WorkerThread @Override public void onLinesListed() {
        onLinesChanged();
        updateHeader();
    }

    @AnyThread @Override public void onLineAdded() {    // todo change to @WorkerThread
        onLinesChanged();
    }

    @WorkerThread @Override public void onPropertiesChanged() {
        updateHeader();
    }

    @WorkerThread @Override public void onBufferClosed() {}

    ////////////////////////////////////////////////////////////////////////////////////////////////

    @MainThread public void loadLinesWithoutAnimation() {
        if (buffer == null) return;
        uiLines.disableAnimationForNextUpdate();
        onLinesChanged();
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    //////////////////////////////////////////////////////////////////////////////////////////////// find hot line
    ////////////////////////////////////////////////////////////////////////////////////////////////

    private final static int HOT_LINE_LOST = -1;
    private final static int HOT_LINE_NOT_PRESENT = -3;

    @MainThread @Cat("Scrolling") public void scrollToHotLineIfNeeded() {
        final int idx = findHotLine();
        if (idx == HOT_LINE_NOT_PRESENT) return;
        if (idx == HOT_LINE_LOST) Weechat.showShortToast(R.string.autoscroll_no_line);
        // run scrolling slightly delayed so that stuff on current thread doesn't get in the way
        else Weechat.runOnMainThread(() -> uiLines.smoothScrollToPositionAfterAnimation(idx), 100);
    }

    @MainThread @Cat(value="Scrolling", exit=true) private int findHotLine() {
        assertThat(buffer).isNotNull();
        assertThat(buffer.linesAreReady()).isTrue();
        final List<Line> lines = _lines;

        int skip = buffer.getHotCount();
        if (skip == 0) return HOT_LINE_NOT_PRESENT;

        Predicate<Line> p = (buffer.type == PRIVATE) ? (l) -> l.type == Line.LINE_MESSAGE :
                (l) -> l.highlighted;

        for (int idx = lines.size() - 1; idx >= 0; idx--)
            if (p.test(lines.get(idx)) && --skip == 0) return idx;

        return HOT_LINE_LOST;
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////////////

    private class DiffCallback extends DiffUtil.Callback {
        private List<Line> oldLines, newLines;

        DiffCallback(List<Line> oldLines, List<Line> newLines) {
            this.oldLines = oldLines;
            this.newLines = newLines;
        }

        @Override public int getOldListSize() {
            return oldLines.size();
        }

        @Override public int getNewListSize() {
            return newLines.size();
        }

        @Override public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
            return oldLines.get(oldItemPosition).pointer == newLines.get(newItemPosition).pointer;
        }

        @Override public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
            return areItemsTheSame(oldItemPosition, newItemPosition);
        }
    }
}

File: app/src/main/java/com/ubergeek42/WeechatAndroid/adapters/ChatLinesAdapter.kt
// Copyright 2012 Keith Johnson
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.ubergeek42.WeechatAndroid.adapters

import android.graphics.PorterDuff
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import androidx.annotation.AnyThread
import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.ubergeek42.WeechatAndroid.R
import com.ubergeek42.WeechatAndroid.R.layout
import com.ubergeek42.WeechatAndroid.Weechat
import com.ubergeek42.WeechatAndroid.copypaste.showCopyDialog
import com.ubergeek42.WeechatAndroid.relay.Buffer
import com.ubergeek42.WeechatAndroid.relay.BufferEye
import com.ubergeek42.WeechatAndroid.relay.HeaderLine
import com.ubergeek42.WeechatAndroid.relay.Line
import com.ubergeek42.WeechatAndroid.relay.LineSpec
import com.ubergeek42.WeechatAndroid.relay.Lines
import com.ubergeek42.WeechatAndroid.relay.MarkerLine
import com.ubergeek42.WeechatAndroid.relay.SquiggleLine
import com.ubergeek42.WeechatAndroid.search.Search
import com.ubergeek42.WeechatAndroid.service.P
import com.ubergeek42.WeechatAndroid.upload.i
import com.ubergeek42.WeechatAndroid.upload.main
import com.ubergeek42.WeechatAndroid.utils.Toaster
import com.ubergeek42.WeechatAndroid.utils.forEachReversedIndexed
import com.ubergeek42.WeechatAndroid.utils.isAnyOf
import com.ubergeek42.WeechatAndroid.utils.ulet
import com.ubergeek42.WeechatAndroid.views.AnimatedRecyclerView
import com.ubergeek42.WeechatAndroid.views.Animation
import com.ubergeek42.WeechatAndroid.views.LineView
import com.ubergeek42.WeechatAndroid.views.solidColor
import com.ubergeek42.WeechatAndroid.views.updateMargins
import com.ubergeek42.cats.Kitty
import com.ubergeek42.cats.Root
import com.ubergeek42.weechat.ColorScheme


class ChatLinesAdapter @MainThread constructor(
    private val uiLines: AnimatedRecyclerView
) : RecyclerView.Adapter<ViewHolder>(), BufferEye {
    @Root private val kitty: Kitty = Kitty.make("ChatLinesAdapter")

    private val inflater = LayoutInflater.from(uiLines.context)

    private var lines = ArrayList<Line>()

    init { setHasStableIds(true) }

    var buffer: Buffer? = null
        @MainThread get
        @MainThread @Synchronized set(value) {
            if (field != value) {
                field = value
                kitty.setPrefix(value?.shortName)
            }
        }

    var onLineDoubleTappedListener: ((Line) -> Unit)? = null

    ////////////////////////////////////////////////////////////////////////////////////////////////
    //////////////////////////////////////////////////////////////////////////////////////// holders
    ////////////////////////////////////////////////////////////////////////////////////////////////


    private inner class Row(view: LineView) : ViewHolder(view) {
        private val lineView = view.apply {
            setOnLongClickListener {
                showCopyDialog(this, buffer?.pointer ?: -1L)
                true
            }

            onDoubleTapListener = {
                (tag as? Line)?.let { line -> onLineDoubleTappedListener?.invoke(line) }
            }
        }

        @MainThread fun update(line: Line) {
            lineView.tag = line
            lineView.setText(line)
        }
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////


    private class ReadMarkerRow(private val view: View) : ViewHolder(view) {
        @MainThread fun update() {
            view.setBackgroundColor(readMarkerColor)
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////


    private class SquiggleRow(private val view: ImageView) : ViewHolder(view) {
        @MainThread fun update() {
            view.setColorFilter(readMarkerColor, PorterDuff.Mode.SRC_IN)
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////


    private inner class HeaderRow(header: View) : ViewHolder(header) {
        private val title: LineView = header.findViewById<LineView>(R.id.title).apply {
            setOnLongClickListener {
                showCopyDialog(this, buffer?.pointer ?: -1)
                true
            }
        }

        private val button: Button = header.findViewById<Button>(R.id.button_more).apply {
            setOnClickListener {
                buffer?.let { buffer ->
                    buffer.requestMoreLines()
                    // instead of calling onLinesListed(), which works,
                    // call the following shortcut to forgo change animation
                    updateButton(buffer.linesStatus)
                }
            }
        }

        @MainThread fun update(line: HeaderLine) {
            updateButton(line.status)
            updateTitle(line)
        }

        @MainThread private fun updateButton(linesStatus: Lines.Status) {
            if (linesStatus === Lines.Status.EverythingFetched) {
                button.visibility = View.GONE
            } else {
                button.visibility = View.VISIBLE
                val canFetchMore = linesStatus === Lines.Status.CanFetchMore
                button.isEnabled = canFetchMore
                button.text = button.context.getString(if (canFetchMore)
                    R.string.ui__button_fetch_more_lines else R.string.ui__button_fetching_lines)
            }
        }

        // don't show the title when fetching lines and only the button is visible --
        // it just doesn't look good when new lines arrive
        @MainThread private fun updateTitle(line: HeaderLine) {
            if (line.spannable.isEmpty() || (itemCount <= 1 && !line.status.ready())) {
                title.visibility = View.GONE
            } else {
                title.visibility = View.VISIBLE
                title.updateMargins(bottom = if (button.visibility == View.GONE) P._4dp.i else 0)
                title.setText(line)
                title.tag = line
            }
        }
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////
    //////////////////////////////////////////////////////////////////////////////////////// adapter
    ////////////////////////////////////////////////////////////////////////////////////////////////


    @MainThread override fun getItemViewType(position: Int): Int {
        return when (lines[position]) {
            is HeaderLine -> HEADER_TYPE
            is MarkerLine -> MARKER_TYPE
            is SquiggleLine -> SQUIGGLE_TYPE
            else -> LINE_TYPE
        }
    }

    @MainThread override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return when (viewType) {
            HEADER_TYPE -> HeaderRow(inflater.inflate(layout.more_button, parent, false))
            MARKER_TYPE -> ReadMarkerRow(inflater.inflate(layout.read_marker, parent, false))
            SQUIGGLE_TYPE -> SquiggleRow(inflater.inflate(layout.squiggle, parent, false) as ImageView)
            else -> Row(LineView(parent.context))
        }
    }

    @MainThread override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        when (val line = lines[position]) {
            is HeaderLine -> (holder as HeaderRow).update(line)
            is MarkerLine -> (holder as ReadMarkerRow).update()
            is SquiggleLine -> (holder as SquiggleRow).update()
            else -> (holder as Row).update(line)
        }
    }

    @MainThread override fun getItemCount() = lines.size

    @MainThread override fun getItemId(position: Int) = lines[position].pointer

    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////////////

    private val updateLock = Object()
    private var updateStep = 0
    private var style = 0

    @AnyThread @Synchronized private fun onLinesChanged(
        animation: Animation = Animation.Default,
        diffLineContents: Boolean = false
    ) = ulet(buffer) { buffer ->
        val thisUpdateStep = synchronized (updateLock) { ++updateStep }

        val newLines = buffer.getLinesCopy()
        val newStyle = buffer.style     // todo synchronization?

        val diffCallback = DiffCallback(lines, newLines, style == newStyle, diffLineContents)
        val diffResult = DiffUtil.calculateDiff(diffCallback, false)

        Weechat.runOnMainThreadASAP {
            synchronized (updateLock) {
                if (thisUpdateStep != updateStep) return@runOnMainThreadASAP
                lines = newLines
                style = newStyle
            }

            diffResult.dispatchUpdatesTo(this@ChatLinesAdapter)

            uiLines.setAnimation(animation)

            if (uiLines.onBottom) {
                uiLines.scrollToPosition(itemCount - 1)
            } else {
                uiLines.flashScrollbar()
            }

            search?.onLinesChanged(newLines)
        }
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////// BufferEye
    ////////////////////////////////////////////////////////////////////////////////////////////////

    @MainThread @Synchronized override fun onGlobalPreferencesChanged(numberChanged: Boolean) {
        onLinesChanged()
    }

    @WorkerThread override fun onLinesListed() {
        onLinesChanged(Animation.NewLinesFetched, diffLineContents = true)
    }

    @AnyThread override fun onLineAdded() {
        onLinesChanged(Animation.LastLineAdded)
    }

    @WorkerThread override fun onTitleChanged() {
        onLinesChanged()
    }

    @WorkerThread override fun onBufferClosed() {}

    ////////////////////////////////////////////////////////////////////////////////////////////////

    @MainThread fun loadLinesWithoutAnimation() {
        onLinesChanged(Animation.None, diffLineContents = true)
    }

    @MainThread @Synchronized fun loadLinesSilently() = ulet(buffer) { buffer ->
        val newLines = buffer.getLinesCopy()
        lines = newLines
    }

    // run scrolling slightly delayed so that stuff on current thread doesn't get in the way
    @MainThread fun scrollToHotLineIfNeeded() {
        when (val idx = findHotLine()) {
            HOT_LINE_NOT_PRESENT -> return
            HOT_LINE_LOST -> Toaster.ShortToast.show(R.string.error__etc__hot_line_lost)
            else -> main(100) { uiLines.smoothScrollToPositionAfterAnimation(idx) }
        }
    }

    private fun findHotLine(): Int {
        var skip = buffer?.hotCount ?: 0
        if (skip == 0) return HOT_LINE_NOT_PRESENT

        lines.forEachReversedIndexed { index, line ->
            if (line.notifyLevel.isAnyOf(LineSpec.NotifyLevel.Highlight,
                                         LineSpec.NotifyLevel.Private)) {
                if (--skip == 0) return index
            }
        }

        return HOT_LINE_LOST
    }

    @MainThread fun findPositionByPointer(pointer: Long): Int {
        lines.forEachIndexed { index, line ->
            if (line.pointer == pointer) return index
        }
        return -1
    }

    var search: Search? = null
        @MainThread set(value) {
            field = value
            value?.onLinesChanged(lines)
        }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    private inner class DiffCallback(
        private val oldLines: List<Line>,
        private val newLines: List<Line>,
        private val sameStyle: Boolean,
        private val diffLineContents: Boolean,
    ) : DiffUtil.Callback() {
        override fun getOldListSize() = oldLines.size
        override fun getNewListSize() = newLines.size

        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return oldLines[oldItemPosition].pointer == newLines[newItemPosition].pointer
        }

        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return if (newItemPosition == 0) {
                // Header line, always present
                oldLines[oldItemPosition] === newLines[newItemPosition]
            } else if (!sameStyle) {
                false
            } else if (diffLineContents) {
                oldLines[oldItemPosition].visuallyEqualsTo(newLines[newItemPosition])
            } else {
                true
            }
        }
    }
}


private const val HEADER_TYPE = -1
private const val LINE_TYPE = 0
private const val MARKER_TYPE = 1
private const val SQUIGGLE_TYPE = 2


private const val HOT_LINE_LOST = -1
private const val HOT_LINE_NOT_PRESENT = -3


private val readMarkerColor get() = ColorScheme.get().chat_read_marker[0].solidColor